index.bs

```diff
@@ -2725,6 +2769,284 @@ The [=remote end steps=] with <var ignore>session</var> and |command parameters|
+|function declaration| and <var ignore>session</var>:
+
+1. Let |returned nodes| be an empty array.
+
+1. For each |context node| of |context nodes|:
+
+   1. Let |script result| be the result of executing the function in |context| defined
+      by the |function declaration| using |context node| as an argument.
+
+   1. If |script result| is an error, return [=error=] with
+      [=error code=] [=invalid selector=].
+   
+   1. If |script result| is not a node list, return [=error=] with
+      [=error code=] [=invalid selector=].
+
+   1. Append |script result| to |returned nodes|.
+
+1. Return |returned nodes|
+
+</div>
+
+<div algorithm="locate a frame context">
+To <dfn>locate a frame context</dfn> with given |context|, |context nodes|
+and |session|:
+
+1. If length of |context nodes| is not equal to 1, return [=error=] with
+   [=error code=] [=invalid selector=].
+
+1. Let |frame node| be the 0th element of |context nodes|.
+
+1. Let |serialization options| be a [=/map=] matching the 
+   <code>script.SerializationOptions</code> production with the fields
+   set to their default values.
+
+1. Let |result ownership| be none.
+
+1. Let |context target| be the a [=/map=] matching of the
+   <code>script.ContextTarget</code> production with the context field
+   set to the [=browsing context id=] of |context|.
+
+1. Let |realm| be the result of [=trying=] to [=get a realm from a target=]
+   given |context target|.
+
+1. Let |window proxy| be the result of [=serialize as a remote value=] with |frame node|,
+   |serialization options|, |result ownership|, a new map as <var ignore>serialization internal map</var>,
+   |realm| and |session|.
+
+1. Return |window proxy|
+
+</div>
+
+<div algorithm="remote end steps for browsingContext.locateNodes">
+The [=remote end steps=] with |session| and |command parameters| are:
+
+1. Let |context id| be the value of the <code>context</code> field of
+   |command parameters|.
+
+1. Let |context| be the result of [=trying=] to [=get a browsing context=]
+   with |context id|.
+
+1. Assert: |context| is not null.
+
+1. Let |locators| be the value of the <code>locators</code> field
+   of |command parameters|.
+
+1. Assert: |locators| is not null.
+
+1. Assert: |locators| length is > 0.
+
+1. Let |start node parameter| be the value of the <code>startNode</code> field
+   of |command parameters|.
+
+1. Let |start node| be the node corresponding to the documentElement of |context|.
+
+1. If |start node parameter| is not null, set |start node| to the result of [=trying=] to
+   [=deserialize local value=] given |start node parameter|, and |session|.
+
+1. Let |wait timeout| be the value of the <code>waitTimeout</code> field of |command
+   parameters| if present, or 0 otherwise.
+
+1. Let |expiration time| be the current time plus |wait timeout| milliseconds.
+
+1. Let |result nodes| be an empty array.
+
+1. Perform the following steps until the current time is greater than |expiration time|
```

Comment:
There's a potential race here: the `expiration time` may have already been reached at this step. We should clarify that these steps must be executed at least once.

Reply:
Absolutely correct. Fixed in subsequent commit.